x-pack/plugins/cloud_security_posture/public/pages/findings/use_findings_by_resource.ts

```diff
@@ -68,31 +82,102 @@ export const getFindingsByResourceAggQuery = ({
   },
 });
 
-export const useFindingsByResource = ({ enabled, index, query }: UseFindingsByResourceOptions) => {
+export const useFindingsByResource = ({
+  enabled,
+  index,
+  query,
+  size,
+  after,
+}: UseFindingsByResourceOptions) => {
   const {
     data,
     notifications: { toasts },
   } = useKibana().services;
 
-  return useQuery(
-    ['csp_findings_resource', { index, query }],
+  const [hasNextPage, setHasNextPage] = useState<boolean>(false);
+  const [afterKeys, setAfterKeys] = useState<FindingsByResourceAggregationsKeys[]>([]);
+
+  const queryResult = useQuery(
+    ['csp_findings_resource', { index, query, size, after }],
     () =>
       lastValueFrom(
         data.search.search<FindingsAggRequest, FindingsAggResponse>({
-          params: getFindingsByResourceAggQuery({ index, query }),
+          params: getFindingsByResourceAggQuery({ index, query, size, after }),
         })
       ),
     {
+      keepPreviousData: true,
+      refetchOnWindowFocus: false,
       enabled,
       select: ({ rawResponse }) => ({
+        after: rawResponse.aggregations?.groupBy?.after_key as
+          | FindingsByResourceAggregationsKeys
+          | undefined,
+        total: rawResponse.hits.total as number,
         page: rawResponse.aggregations?.groupBy.buckets.map(createFindingsByResource) || [],
       }),
       onError: (err) => showErrorToast(toasts, err),
     }
   );
+
+  const getNextKey = () => queryResult.data?.after;
+
+  const getPreviousKey = () => {
+    // when length is 0, we're on the 1st page, so no previous key
+    // when length is 1, we're on the 2nd page, so previous key is 'undefined'
+    if (afterKeys.length < 2) return undefined;
+
+    // We need the 'after_key' from the page that points to the previous page
+    return afterKeys[afterKeys.length - 2];
+  };
+
+  /**
+   * Keeps track of pages we can go back to
+   */
+  useEffect(() => {
+    if (!enabled) return;
+    setAfterKeys((keys) => {
+      if (!after) return [];
+      const hasKeyBeenSeen = keys.some((key) => isEqual(key, after));
+
+      return hasKeyBeenSeen
+        ? keys.slice(0, -1) // We go back one at a time
+        : [...keys, after];
+    });
```

Comment:
this bit makes sure we keep track of an array of `after_keys`  by adding the next key to the array (if it's new) or removing it (if it's been seen)